synapse/handlers/receipts.py

```diff
@@ -116,6 +116,14 @@ async def _handle_new_receipts(self, receipts: List[ReadReceipt]) -> bool:
         min_batch_id: Optional[int] = None
         max_batch_id: Optional[int] = None
 
+        # Beeper: we don't want to send read receipts to large rooms,
+        # so we convert messages to private, that are over RECEIPT_MAX_ROOM_SIZE.
+        for i, r in enumerate(receipts):
+            if r.receipt_type != ReceiptTypes.READ_PRIVATE:
+                num_users = await self.store.get_number_joined_users_in_room(r.room_id)
```

Comment:
Should we gather all of these ahead of time? Also a little clunky awaited on these one at a time, can we bulk grab these with twisted async somehow? @Fizzadar plz

Reply:
Note this handles federation as well so there could be up to 100 read receipts here

Reply:
updated to ensure we only check the size of each room once https://github.com/beeper/synapse/pull/67/commits/fbcd582f6e988b7ac6176807ab2b70e91889b8f2

But would be great to hear if anyone knows how to do the equivalent of `asyncio.gather` in twisted, I haven't found it myself yet....

Reply:
There’s a [`gatherResults`](https://twisted.org/documents/13.1.0/api/twisted.internet.defer.gatherResults.html) that roughly does the same!

Reply:
Nice, I'll check that out when I next get a chance to work on this.